src/v/storage/tests/log_segment_appender_test.cc

```diff
@@ -212,3 +217,24 @@ SEASTAR_THREAD_TEST_CASE(test_can_append_little_data) {
     BOOST_REQUIRE_EQUAL(appender.file_byte_offset(), data.size());
     appender.close().get();
 }
+
+SEASTAR_THREAD_TEST_CASE(test_full_disk_append_failure) {
+    auto f = ss::file(
+      ss::make_shared(fs_finject(open_file("test_segment_appender.log"))));
+    auto appender = make_segment_appender(f);
+    auto badger = finjector::shard_local_badger();
+    badger.register_probe(fs_failure_probe::name(), &fs_finject::get_probe());
+    badger.set_exception(fs_failure_probe::name(), "allocate");
+    badger.set_exception(fs_failure_probe::name(), "write_dma");
+
+    constexpr size_t one_meg = 1024 * 1024;
+    iobuf original = make_random_data(one_meg);
+    BOOST_REQUIRE_THROW(
+      {
+          appender.append(original).get();
+          appender.flush().get();
+          appender.close().get();
```

Comment:
This close() doesn't get called, obviously, but I get a failed assertion when the destructor deletes it, but it hasn't been closed. I thought "RAII means we don't need finally in C++", so I was a little surprised. I'll update this code shortly to move the flush/close below this block.

Reply:
done.